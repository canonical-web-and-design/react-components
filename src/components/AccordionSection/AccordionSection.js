import PropTypes from "prop-types";
import React, { useRef } from "react";
import uuidv4 from "uuid/v4";

const AccordionSection = ({
  content,
  expanded,
  onTitleClick,
  setExpanded,
  title
}) => {
  const buttonId = useRef(uuidv4());
  const sectionId = useRef(uuidv4());
  const isExpanded = expanded === buttonId.current;
  return (
    <li className="p-accordion__group">
      <button
        aria-controls={`#${sectionId.current}`}
        aria-expanded={isExpanded ? "true" : "false"}
        className="p-accordion__tab"
        id={buttonId.current}
        onClick={() => {
          if (isExpanded) {
            setExpanded(null, null);
          } else {
            setExpanded(buttonId.current, title);
          }
          onTitleClick && onTitleClick(!isExpanded);
        }}
        role="tab"
        type="button"
      >
        {title}
      </button>
      <section
        aria-hidden={isExpanded ? "false" : "true"}
        aria-labelledby={sectionId.current}
        className="p-accordion__panel"
        id={sectionId.current}
        role="tabpanel"
      >
        {content}
      </section>
    </li>
  );
};

AccordionSection.propTypes = {
  content: PropTypes.node,
  expanded: PropTypes.string,
  /**
   * An optional click event when the title is clicked.
   */
  onTitleClick: PropTypes.func,
  setExpanded: PropTypes.func,
  title: PropTypes.string
};

export default AccordionSection;
