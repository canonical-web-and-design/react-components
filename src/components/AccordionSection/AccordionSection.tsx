import PropTypes from "prop-types";
import React, { useRef } from "react";
import type { ElementType, ReactNode } from "react";
import uuidv4 from "uuid/v4";

type Props = {
  content?: ReactNode;
  expanded?: string;
  onTitleClick?: (expanded: boolean, key: string) => void;
  sectionKey?: string;
  setExpanded?: (key: string | null, title: string | null) => void;
  title?: string;
  titleElement?: ElementType | null;
};

const AccordionSection = ({
  content,
  expanded,
  onTitleClick,
  sectionKey,
  setExpanded,
  title,
  titleElement,
}: Props): JSX.Element => {
  const sectionId = useRef(uuidv4());
  const key = sectionKey || sectionId.current;
  const isExpanded = expanded === key;
  const Title = titleElement || null;
  const hasTitleElement = !!Title;

  return (
    <li className="p-accordion__group">
      <button
        aria-controls={`#${sectionId.current}`}
        aria-expanded={isExpanded ? "true" : "false"}
        className={
          hasTitleElement ? "p-accordion__tab--with-title" : "p-accordion__tab"
        }
        onClick={() => {
          if (isExpanded) {
            setExpanded(null, null);
          } else {
            setExpanded(key, title);
          }
          onTitleClick && onTitleClick(!isExpanded, key);
        }}
        role="tab"
        type="button"
      >
        {hasTitleElement ? (
          <Title className="p-accordion__title">{title}</Title>
        ) : (
          title
        )}
      </button>
      <section
        aria-hidden={isExpanded ? "false" : "true"}
        aria-labelledby={sectionId.current}
        className="p-accordion__panel"
        id={sectionId.current}
        role="tabpanel"
      >
        {content}
      </section>
    </li>
  );
};

AccordionSection.propTypes = {
  content: PropTypes.node,
  expanded: PropTypes.string,
  /**
   * An optional click event when the title is clicked.
   */
  onTitleClick: PropTypes.func,
  /**
   * An optional key to be used to track which section is selected.
   */
  sectionKey: PropTypes.string,
  setExpanded: PropTypes.func,
  title: PropTypes.string,
  titleElement: PropTypes.oneOf(["h2", "h3", "h4", "h5", "h6"]),
};

export default AccordionSection;
