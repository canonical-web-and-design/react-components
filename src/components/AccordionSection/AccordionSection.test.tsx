import { shallow } from "enzyme";
import React from "react";

import AccordionSection from "./AccordionSection";

const MOCK_UUID = "00000000-0000-0000-0000-000000000000";

jest.mock("uuid/v4", () => jest.fn(() => MOCK_UUID));

describe("AccordionSection ", () => {
  it("renders", () => {
    const wrapper = shallow(
      <AccordionSection
        content={<span>Test</span>}
        expanded="abcd-1234"
        setExpanded={jest.fn()}
        title="Test section"
      />
    );

    expect(wrapper).toMatchSnapshot();
  });

  it("renders headings for titles", () => {
    const wrapper = shallow(
      <AccordionSection
        content={<span>Test</span>}
        expanded="abcd-1234"
        setExpanded={jest.fn()}
        title="Test section"
        titleElement="h4"
      />
    );

    expect(wrapper.find("h4.p-accordion__title")).toHaveLength(1);
    expect(wrapper).toMatchSnapshot();
  });

  it("can handle click events on the title", () => {
    const onTitleClick = jest.fn();
    let expanded = null;
    const wrapper = shallow(
      <AccordionSection
        content={<span>Test</span>}
        expanded={expanded}
        onTitleClick={onTitleClick}
        setExpanded={(id) => {
          expanded = id;
        }}
        title="Test section"
      />
    );
    wrapper.find(".p-accordion__tab").at(0).simulate("click");

    expect(onTitleClick.mock.calls[0]).toEqual([true, MOCK_UUID]);
    wrapper.setProps({ expanded });
    // Clicking the title again should close the accordion section.
    wrapper.find(".p-accordion__tab").at(0).simulate("click");

    expect(onTitleClick.mock.calls[1]).toEqual([false, MOCK_UUID]);
  });

  it("can use a key for expanded state", () => {
    const onTitleClick = jest.fn();
    let expanded = null;
    const wrapper = shallow(
      <AccordionSection
        content={<span>Test</span>}
        expanded={expanded}
        sectionKey="first-key"
        onTitleClick={onTitleClick}
        setExpanded={(id) => {
          expanded = id;
        }}
        title="Test section"
      />
    );
    wrapper.find(".p-accordion__tab").at(0).simulate("click");

    expect(onTitleClick.mock.calls[0]).toEqual([true, "first-key"]);
  });
});
