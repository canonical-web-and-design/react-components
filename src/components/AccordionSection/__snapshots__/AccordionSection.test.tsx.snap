// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccordionSection  renders 1`] = `
<li
  className="p-accordion__group"
>
  <button
    aria-controls="#Uakgb_J5m9g-0JDMbcJqLJ"
    aria-expanded="false"
    className="p-accordion__tab"
    onClick={[Function]}
    role="tab"
    type="button"
  >
    Test section
  </button>
  <section
    aria-hidden="true"
    aria-labelledby="Uakgb_J5m9g-0JDMbcJqLJ"
    className="p-accordion__panel"
    id="Uakgb_J5m9g-0JDMbcJqLJ"
    role="tabpanel"
  >
    <span>
      Test
    </span>
  </section>
</li>
`;

exports[`AccordionSection  renders headings for titles 1`] = `
<li
  className="p-accordion__group"
>
  <button
    aria-controls="#Uakgb_J5m9g-0JDMbcJqLJ"
    aria-expanded="false"
    className="p-accordion__tab--with-title"
    onClick={[Function]}
    role="tab"
    type="button"
  >
    <h4
      className="p-accordion__title"
    >
      Test section
    </h4>
  </button>
  <section
    aria-hidden="true"
    aria-labelledby="Uakgb_J5m9g-0JDMbcJqLJ"
    className="p-accordion__panel"
    id="Uakgb_J5m9g-0JDMbcJqLJ"
    role="tabpanel"
  >
    <span>
      Test
    </span>
  </section>
</li>
`;
